stonesoup/types/track.py

```diff
@@ -42,11 +57,33 @@ def metadata(self):
             All metadata associate with this track.
         """
 
-        metadata = {}
+        return self._metadata
 
-        for state in self:
-            if isinstance(state, Update) \
-                    and state.hypothesis.measurement.metadata is not None:
-                metadata.update(state.hypothesis.measurement.metadata)
+    def _update_metadata_from_state(self, state):
+        """ Extract and update track metadata, given a state
+
+        Parameters
+        ----------
+        state: State
+            A state object from which to extract metadata. Metadata can only
+            be extracted from Update (or subclassed) objects. Calling this
+            method with a non-Update (subclass) object will NOT return an
+            error, but will have no effect on the metadata.
+
+        """
 
-        return metadata
+        if isinstance(state, Update):
+            if isinstance(state.hypothesis, MultipleHypothesis):
+                # Sort and iterate through multiple hypotheses such that most
+                # likely hypothesis comes last. This ensures that metadata
+                # from all hypotheses are retained, but more likely
+                # hypotheses will over-write the metadata set by less likely
+                # ones.
+                for hypothesis in sorted(state.hypothesis, reverse=True):
```

Comment:
Not sure if this is correct. My (possibly wrong) intuition says that calling `sorted` on a MultipleHypothesis object should return a list of hypotheses, with the most likely being first. In the case of `SingleProbabilityHypothesis` this means that the hypothesis with highest probability would come first, while for `SingleDistanceHypothesis` the hypothesis with lowest distance would be first.. Thus, to have them sorted in reverse, I have included the argument `reverse=True`.